File: app/src/main/java/remix/myplayer/ui/widget/VerticalScrollTextView.java
package remix.myplayer.ui.widget;

import android.content.Context;
import androidx.annotation.Nullable;
import androidx.annotation.StringRes;
import androidx.appcompat.widget.AppCompatTextView;
import android.util.AttributeSet;

/**
 * 滚动换行
 */
public class VerticalScrollTextView extends AppCompatTextView {

  public static final int VERTICAL = 0;
  public static final int HORIZONTAL = 1;

  private static final int DURATION = 2000;

  private int mOrientation = VERTICAL;


  public VerticalScrollTextView(Context context) {
    super(context);
  }

  public VerticalScrollTextView(Context context, @Nullable AttributeSet attrs) {
    super(context, attrs);
  }

  public VerticalScrollTextView(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
    super(context, attrs, defStyleAttr);
  }

  public void setOrientation(int orientation) {
    if (orientation != HORIZONTAL && orientation != VERTICAL) {
      throw new IllegalArgumentException("Invalid orientation");
    }
    mOrientation = orientation;
  }

  public void setTextWithAnimation(@StringRes int textRes) {
    setTextWithAnimation(getResources().getString(textRes));
  }

  public void setTextWithAnimation(CharSequence text) {
    setText(text);
  }
}

File: app/src/main/java/remix/myplayer/ui/widget/VerticalScrollTextView.kt
package remix.myplayer.ui.widget

import android.content.Context
import android.util.AttributeSet
import androidx.annotation.StringRes
import androidx.appcompat.widget.AppCompatTextView

/**
 * 滚动换行
 */
class VerticalScrollTextView : AppCompatTextView {
  private var orientation = VERTICAL

  constructor(context: Context) : super(context) {}
  constructor(context: Context, attrs: AttributeSet?) : super(context, attrs) {}
  constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {}

  fun setOrientation(orientation: Int) {
    require(!(orientation != HORIZONTAL && orientation != VERTICAL)) { "Invalid orientation" }
    this.orientation = orientation
  }

  fun setTextWithAnimation(@StringRes textRes: Int) {
    setTextWithAnimation(resources.getString(textRes))
  }

  fun setTextWithAnimation(text: CharSequence?) {
    setText(text)
  }

  companion object {
    const val VERTICAL = 0
    const val HORIZONTAL = 1
    private const val DURATION = 2000
  }
}